prov/efa/src/rxr/rxr_ep.c

```diff
@@ -1065,44 +1065,21 @@ static ssize_t rxr_ep_cancel(fid_t fid_ep, void *context)
  */
 static int efa_set_fi_hmem_p2p_opt(struct efa_ep *efa_ep, int opt)
 {
-	struct efa_hmem_info *hmem_info;
+	int i, ret;
 
-	hmem_info = efa_ep->domain->hmem_info;
-
-	switch (opt) {
 	/*
-	 * TODO: support the other options. We can only support ENABLED
-	 * and PREFERRED when p2p is available. DISABLED is not
-	 * supported yet.
+	 * Check the opt's validity against the first initialized non-system FI_HMEM
+	 * interface
 	 */
-	case FI_HMEM_P2P_REQUIRED:
-	case FI_HMEM_P2P_ENABLED:
-	case FI_HMEM_P2P_PREFERRED:
-		if (hmem_info[FI_HMEM_CUDA].initialized &&
-		    hmem_info[FI_HMEM_CUDA].p2p_supported_by_device) {
-			efa_ep->hmem_p2p_opt = opt;
-		} else if (hmem_info[FI_HMEM_NEURON].initialized &&
-			   hmem_info[FI_HMEM_NEURON].p2p_supported_by_device) {
-			/*
-			 * Neuron requires p2p support and supports no
-			 * other modes.
-			 */
-			if (opt != FI_HMEM_P2P_REQUIRED)
-				return -FI_EOPNOTSUPP;
-			efa_ep->hmem_p2p_opt = FI_HMEM_P2P_REQUIRED;
-		} else {
-			return -FI_EOPNOTSUPP;
+	EFA_HMEM_IFACE_FOREACH_NON_SYSTEM(i) {
+		ret = efa_hmem_validate_p2p_opt(efa_ep->domain, efa_hmem_ifaces[i], opt);
```

Comment:
the logic is a bit hard to understand. Also `efa_ep->hmem_p2p_opt` should be updated after we go through all interfaces.

How about

```
EFA_HMEM_IFACE_FOREACH_NON_SYSTEM(i) {
err = efa_hmem_validate_p2p_opt(efa_ep->domain, efa_hmem_ifaces[i], opt);
if (err == -FI_ENODATA)
    continue

if (err)
   return err;
}

efa_ep->hmem_p2p_opt  = opt;
```

Reply:
> after we go through all interfaces

This would diverge from our current behavior. We currently iterate through until we encounter the first initialized interface and break. Validating the p2p option against *every* initialized interface would be new logic. Is this what we're doing going forward?

Reply:
Yes. I think so. Current behavior is written with single interface in mind. Multiple interface logic should like this